Sub-stakes description (#1558)
nucypher/nucypher: docs/source/architecture/sub_stakes.rst

Sub-stakes
==========

A staker may extend the unlock period for any number of portions of their total stake. This divides the stake into smaller parts, each with a unique unlock date in the future. Stakers may also acquire and lock new tokens. The total stake is represented as the sum of all the different sub-stakes active in a given cycle (new cycle every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn. Each sub-stake has a beginning and duration (lock time). When a staker confirms activity each day and winding down is enabled, the remaining lock time for relevant sub-stakes is reduced.

> Sub-stake remains active until become unlocked and staker get reward for last period of sub-stake by calling ``mint()`` or ``confirmActivity()`` after the last period is surpassed. Each staker can have no more than 30 active sub-stakes. All sub-stakes changes use slots of inactive sub-stakes in the first place and only after that - empty slots in the array. Therefore, obtaining data on past staker locks is not guaranteed.

Author: Few moments:
- Collecting reward is not necessary to make sub-stake inactive. We can "remove" element from sub-stakes array when we no need information about locking. After last period of sub stake become history we still need to calculate reward for this last period. It will happen inside mint() method wich work only with surpassed period.
- Not sure about "the inactive stake slots", for me sounds like slot is inactive (maybe I'm wrong). I'll try to rephrase idea: we have array with sub-stakes with fixed size (30). Some of this sub-stakes are not usefull for any calculation in StakingEscrow, they are inactive and we can reuse them. When we add new one - we just looking from left to right trying to find place where to save new element, inactive sub-stake or empty slots if there are any. Not sure that is more clear than before :)
- We allways can get information about locking for any staker for the current moment and for future, also we could try to retreive this information for some period in the past, and if were no any rewrites of inactive slots then this information still will be valid

> Sub-stake remains active until become unlocked and staker get reward for last period of sub-stake by calling ``mint()`` or ``confirmActivity()`` after the last period is surpassed. Each staker can have no more than 30 active sub-stakes. All sub-stakes changes use slots of inactive sub-stakes in the first place and only after that - empty slots in the array. Therefore, obtaining data on past staker locks is not guaranteed.

Reviewer: Thanks for the explanation - that makes things clearer. I guess the terminology `staker locks` is just a little weird to me.

```suggestion
A sub-stake remains active until it becomes unlocked, and a staker collects the reward for the last period by calling ``mint()`` or ``confirmActivity()`` once the last period is surpassed. Each staker can have no more than 30 active sub-stakes which are stored in an array. All sub-stake changes initially reuse slots of inactive sub-stakes for storage in the array, and if there are none, will instead use empty slots. Therefore, attempting to retrieve data about previous inactive sub-stakes is not guaranteed to be successful since the data could have been overwritten.
```

> Sub-stake remains active until become unlocked and staker get reward for last period of sub-stake by calling ``mint()`` or ``confirmActivity()`` after the last period is surpassed. Each staker can have no more than 30 active sub-stakes. All sub-stakes changes use slots of inactive sub-stakes in the first place and only after that - empty slots in the array. Therefore, obtaining data on past staker locks is not guaranteed.

Author: Great! Just changed `collects` -> `gets`, because we have CLI command that collects reward and this command has other meaning

> A sub-stake remains active until it becomes unlocked, and a staker gets the reward for the last period by calling ``mint()`` or ``confirmActivity()`` once the last period is surpassed. Each staker can have no more than 30 active sub-stakes which are stored in an array. All sub-stake changes initially reuse slots of inactive sub-stakes for storage in the array, and if there are none, will instead use empty slots. Therefore, attempting to retrieve data about previous inactive sub-stakes is not guaranteed to be successful since the data could have been overwritten.

Author: Can't accept this suggestion, some little things do changes in emphasize. Splitting first sentence into two makes other meaning that I want: active sub-stake is two condition but in suggetion looks like only one.
And don't want to put info about empty slots to parentheses

Operations that modify the sub-stake array
------------------------------------------

Deposit and locking
^^^^^^^^^^^^^^^^^^^
*Methods* : ``deposit(uint256,uint16)``,  ``deposit(address,uint256,uint16)``,  ``lock(uint256,uint16)``

> To become a staker, tokens owner transfers tokens to ``StakingEscrow`` and lock them  using one of the ``deposit()`` methods. In case that staker already has some amount of unlocked tokens on the account in the contract then lock can be created using ``lock()`` method. As a result of executing on of the above methods, a new element will be created in the array of sub-stakes with the next period as starting date  and with duration equals input parameter.

Reviewer: ```suggestion
To become a staker, NU tokens must be transferred to the ``StakingEscrow`` contract and locked using one of the ``deposit()`` methods. If the staker already has unlocked tokens within the account in the contract, then the sub-stake can be locked using the ``lock()`` method. If successful, a new element will be created in the array of sub-stakes with the next period as the starting date, and the duration equal to the input parameter.
```

Author: I like sentence, just not sure about  `the sub-stake can be locked`, in my head it sounds like sub-stake can be only created and locking tokens is staking/creating satake. What do you think?

Reviewer: I saw your update: 👍 .

**Example:**

    A staker deposits 900 tokens:
        * 1st sub-stake = 900 tokens starting from the next period and locked for 5 periods

    .. aafig::
        :proportional:
        :textual:

            stake
            ^
            |
         900|   +-------------+
            |   |     1st     |  period
            +---+-------------+--->
            0   1             5

Lock prolongation
^^^^^^^^^^^^^^^^^
*Methods* : ``prolongStake(uint256,uint16)``

> In order to increase the staking reward, as well as the possibility of obtaining policies with a longer time,  staker can increase the locking duration of the sub-stake using the ``prolongStake()`` method. The number of sub-stakes does not change, but the end period of locking in the specified sub-stake will be changed.

Reviewer: ```suggestion
In order to increase the staking reward, as well as the possibility of obtaining policies with a longer timeframe, stakers can increase the duration of their locked sub-stake using the ``prolongStake()`` method. The number of sub-stakes does not change, but the locked duration for the specified sub-stake will be increased.
```

Author: ✔️

**Example:**

    A staker prolongs sub-stake for 2 additional periods up to 7:
		- Before: 
			* 1st sub-stake = 900 tokens for 5 periods
		- After: 
			* 1st sub-stake = 900 tokens for 7 periods

    .. aafig::
        :proportional:
        :textual:

                 Before             

            stake                 
            ^                     
            |                     
         900+----------+          
            |    1st   |  period
            +----------+--->   
            0          5          
			
			
			
                 After             

            stake                 
            ^                     
            |                     
         900+-------------+        
            |      1st    |  period
            +-------------+--->   
            0             7     
			
			
Sub-stake division
^^^^^^^^^^^^^^^^^^
*Methods* : ``divideStake(uint256,uint256,uint16)``

> If necessary, staker can extend the locking duration only for the part of tokens in a sub-stake (``divideStake()``). Contract method splits desirable sub-stake into two elements: the old sub-stake with changed locked amount and the new sub-stake. New element has the extended locking duration (by the specified number of periods) and the same start period as the old one has. Locked amount in the new sub-stake equals to requested input parameter.

Reviewer: ```suggestion
If necessary, stakers can extend the locking duration for only a portion of their tokens in a sub-stake by using the ``divideStake()`` method. This method splits a sub-stake into two elements: the old sub-stake with the reduced locked amount and the new sub-stake with the specified amount. The new sub-stake has the specified locked amount and an extended lock duration, based on the specified number of periods, with the same start period as the old sub-stake.
```

Author: ✔️

		- After: 
			* 1st sub-stake = 600 tokens for 5 periods
			* 2nd sub-stake = 300 tokens for 7 periods

    .. aafig::
        :proportional:
        :textual:

                 Before             

            stake                 
            ^                     
            |                     
         900+----------+          
            |          |  
            |    1st   |
            |          |  period 
            +----------+--->   
            0          5          
			
			
			
                 After             

            stake                 
            ^                     
            |                     
         900+----------+       
            |          |
            |    1st   |
         300+----------+--+
            |      2nd    |  period
            +-------------+--->   
            0          5  7     

Slashing
^^^^^^^^
*See:* `the slashing protocol`_ 

.. _`the slashing protocol`: https://docs.nucypher.com/en/latest/architecture/slashing.html

Flags that affect the sub-stake array
-------------------------------------

Re-staking
^^^^^^^^^^
*Used in methods* : ``confirmActivity()``, ``mint()``

> When re-staking is turned off, the number of locked tokens in sub-stakes does not change by itself.

Reviewer: Should we indicate what the default is? - like we did for `wind-down`

Author: Thanks, I forgot to add this

> In case when re-staking parameter is on then all staking reward is locked as the part of each involved sub-stake (inside ``confirmActivity()`` and/or ``mint()``).  Accordingly, each such sub-stake has increased locked amount (by reward) and the number of sub-stakes stays unchanged.

Reviewer: ```suggestion
However, when re-staking is enabled then all staking rewards are re-locked as part of each relevant sub-stake (inside ``confirmActivity()`` and/or ``mint()``).  Consequently, each such sub-stake has an increased locked amount (by reward) and the number of sub-stakes remains unchanged.
```

Author: ✔️

**Example:**

>     A staker has few sub-stakes and calls ``mint()``:

Reviewer: For this scenario, is it the case that 1st sub-stake produces 50 tokens and 2nd sub-stake produces 20 tokens as rewards? If so, we should state that explicitly for the scenario.

Author: Yep, you're totally right, where to put it? in description of before? Or in "after" sections?

Author: I took your suggestion with some modifications, tell me what do you think

Reviewer: 👍 added minor comment about tense.

>     A staker has few sub-stakes and calls ``mint()``. Assume that thus far the 1st and 2nd sub-stakes will produced 50 tokens and 20 tokens respectively in rewards:

Reviewer: You can either do `will produce` (future tense) or `have produced` (past tense)

Author: Thanks, I tried to use future tense but failed :)

		- Before calling: 
			* 1st sub-stake = 400 tokens for 10 periods
			* 2nd sub-stake = 200 tokens for 2 periods
			* 3rd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods
			* 100 tokens in an unlocked state

> 		- After calling, re-staking is disabled:  

Author: Agreed, I like it
✔️

			* 1st sub-stake = 400 tokens for 10 periods
			* 2nd sub-stake = 200 tokens for 2 periods
			* 3rd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods
			* 170 tokens in an unlocked state
		- After calling, re-staking is enabled: 
			* 1st sub-stake = 450 tokens for 10 periods
			* 2nd sub-stake = 220 tokens for 2 periods
			* 3rd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods
			* 100 tokens in an unlocked state

    .. aafig::
        :proportional:
        :textual:

                 Before             

            stake
            ^
            |
         700|     +----+
            |     | 3rd|
         600+-----+----+
         500|          +-------------+
            |    2nd   |     3rd     |
         400+----------+-------------+----------+
            |                                   |
            |               1st                 |   period
            +-----------------------------------+--->    

>             0     1    2             5          10

Reviewer: Since the `3rd` sub-stake is to start at period and last for 5 periods, should the end of the `3rd` sub-stake correspond to `6` instead of `5` on the timeline? - Same for the "After" diagram below?

Author: I'm glad that somebody asked this :)
So thing is that locking for 5 periods means 5 future period - 1,2,3,4,5.  3rd sub-stake will last up to 5th period and will end in 6th. Current period is already part of history from this perspective.

Author: I tried to show discrete figure, when there is no such thing like 1->2, for example lock for one period  will be shown as vertical line. 
In all calculation current period ignored, so almost everything starts from 1 on those graphs. So lock for 10 periods will be for 1,2,3...9,10 period number, 0 is not included in this row. I guess this's problem in the explanation, maybe add "future periods" everywhere?

Author: How about such figure:
![image](https://user-images.githubusercontent.com/22225795/72432807-ff91cc80-37a8-11ea-929e-8abd331f2677.png)
I'm trying to use `aafigure` lib, it helps to maintain in future, but if still not good - I'll search something else

Reviewer: @vzotova, 👍 I can get on board with that if that works with `aafigure`.

Author: Updated all diagrams 
waiting your review (also better to check version after compilation)

			
			
                 After, re–staking is enabled             

            stake
            ^
            |
         770|     +----+
            |     | 3rd|
         670+-----+----+
            |          |
         550|    2nd   +-------------+
            |          |     3rd     |
         450+----------+-------------+----------+
            |                                   |
            |               1st                 |
            |                                   |   period
            +-----------------------------------+--->    
            0     1    2             5          10
			

Winding down
^^^^^^^^^^^^
*Used in methods* : ``confirmActivity()``

> Disabled winding down parameter (by default) guarantees that worker must call ``confirmActivity()`` at least N times after parameter will be turned on to unlock sub-stake, where N is locking duration of sub-stake. Thus unlocking date for each sub-stakes shifts by 1 period each period (duration remains the same). In case when winding down is enabled, each ``confirmActivity()`` (no more than once in a period) leads to decreasing of locking duration of each sub-stake. If worker calls ``confirmActivity()`` each period then unlocking date remains unchanged.

Reviewer: ```suggestion
A disabled "winding down" parameter (default) guarantees that the worker must call ``confirmActivity()`` at least N times after the parameter is enabled to unlock the sub-stake, where N is the locking duration of sub-stake. When disabled, the unlock date for each sub-stakes shifts forward by 1 period after each period i.e. the duration continues to remain the same until the parameter is enabled. Once the "winding down" parameter is enabled, each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of locking duration for each sub-stake, and the unlock date no longer changes.
```

Author: ✔️

Sub-stakes
==========

> A staker may extend the unlock period for any number of portions of their total stake. This divides the stake into smaller parts, each with a unique unlock date in the future. Stakers may also acquire and lock new tokens. The total stake is represented as the sum of all the different sub-stakes active in a given cycle (new cycle every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn. Each sub-stake has a beginning and duration (lock time). When a staker confirms activity each day and winding down is enabled, the remaining lock time for relevant sub-stakes is reduced.

Reviewer 2: Should we use consistent terminology? Here we use `cycle` - is that the same thing as a `period`?

Author: Changed to `period`
✔️

Operations that modify the sub-stake array
------------------------------------------

Deposit and locking
^^^^^^^^^^^^^^^^^^^
*Methods* : ``deposit(uint256,uint16)``,  ``deposit(address,uint256,uint16)``,  ``lock(uint256,uint16)``

To become a staker, NU tokens must be transferred to the ``StakingEscrow`` contract and locked using one of the ``deposit()`` methods. If the staker already has unlocked tokens within the account in the contract, then the stake can be locked using the ``lock()`` method. If successful, a new element will be created in the array of sub-stakes with the next period as the starting date, and the duration equal to the input parameter.

**Example:**

    A staker deposits 900 tokens:
        * 1st sub-stake = 900 tokens starting from the next period and locked for 5 periods

    .. aafig::
        :proportional:
        :textual:

            stake
            ^
            |
         900|   +-------------------+
            |   |        1st        |   period
            +---+---+---+---+---+---+---->

>             | 0 | 1 | 2 | 3 | 4 | 5 | 6

Reviewer: 👍

Lock prolongation
^^^^^^^^^^^^^^^^^
*Methods* : ``prolongStake(uint256,uint16)``

In order to increase the staking reward, as well as the possibility of obtaining policies with a longer timeframe, stakers can increase the duration of their locked sub-stake using the ``prolongStake()`` method. The number of sub-stakes does not change, but the locked duration for the specified sub-stake will be increased.

**Example:**

    A staker prolongs sub-stake for 2 additional periods up to 7:
		- Before: 
			* 1st sub-stake = 900 tokens for 5 periods
		- After: 
			* 1st sub-stake = 900 tokens for 7 periods

    .. aafig::
        :proportional:
        :textual:

                         Before             

            stake
            ^
            |
         900+-----------------------+
            |          1st          |   period
            +---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6      

> 			

Reviewer: @vzotova This is where I get confused - (and apologies, if I'm just totally missing it) - but shouldn't the first sub-stake for the `before` scenario stop at the end of period 4, and not at the end of period 5? i.e. the 1st sub-stake includes the current period and is for periods 0, 1, 2, 3, 4 (total = 5 periods). 

This stake starts from the current period (0), no? or is it 5 periods starting the next period, like in the `Deposit and locking` scenario above, in which case it should start from period 1?

Author: current period does not count, it's sort of historical data, for example lock for 0 periods looks like
```
^
|
+---+
|   |
+---+--->
| 0 | 1
```

Reviewer: Perhaps it boils down to a wording issue - we should indicate whether the current period is included or not, eg.

```
* 1st sub-stake = 900 tokens for 5 periods (in addition to the current period)
```

wdyt? Open to alternative wording as well.

@arjunhassard wdyt about the wording?

Author: I'm for describing this somewhere in the first section, because current period always does not count when we talk about locking duration

Reviewer: 👍 - Adding it as an overall statement makes sense.

Reviewer 2: I agree that it'd be nice for some additional clarity on this. You can see from my other comments that I'm quite confused about it 😆

Author: I'm a little bit stuck with formulating this, could you help me
What I want to say:
* duration (phrases "locked for ") means how many times need to call `confirmActivity()` so this leads to thought that duration is only future periods (current does not count)
* locking in the current period depends on waht situation was in past - if we step 1 period to the previous period and look into next one (which is current period in the current moment), if next period is locked for some sub-stake that means lock in the current period in the current moment and vice versa

Reviewer: So just to ensure that I understand, we define `locking duration` as the number of times `confirmActivity()` needs to be called.

If a stake has a locking duration of 5 periods it means that the stake will be active for periods: 0,1,2,3,4,5 because `confirmActivity()` would be called during periods 0,1,2,3,4 each time confirming activity for the subsequent period (1,2,3,4,5) respectively.

Is this right?

Author: almost perfect, 0th period lock is not guaranteed, so lock will be active for periods: 1,2,3,4,5 and 0th period should be specified additionally (which by default is locked and unlocked when said that lock starts in the next period)

Author: summing up, the correct explanation:
`If a stake has a locking duration of 5 periods it means that the stake will be active for periods: 1,2,3,4,5 because confirmActivity() would be called during periods 0,1,2,3,4 each time confirming activity for the subsequent period (1,2,3,4,5) respectively.`

Reviewer: Right, so unless specified otherwise (example by stating "*starting from the next period *") the 0th period is considered active.

---

@vzotova what about something like the following at the top of the doc (after the sentence that ends "and can be freely withdrawn.").

```
[and can be freely withdrawn.]

Each sub-stake has a beginning and locking duration. For our terminology, we define `locking duration` to be the required number of times that `confirmActivity()` needs to be called so that the stake is registered as active for the subsequent period. For example, if a stake has a *locking duration of 5 periods* it means that the stake will be active for periods: 1,2,3,4,5 because `confirmActivity()` would be called during periods 0,1,2,3,4 each time confirming activity for the subsequent period (1,2,3,4,5) respectively. Unless the starting period is specified, the stake is considered active for the 0th (current) period by default.
```

Obviously you can reword as needed.

Reviewer: I was also thinking we could remove the sentence at the top (Overview):

` When a staker confirms activity each day and winding down is enabled, the remaining lock time for relevant sub-stakes is reduced.` 

since wind-down will be explained appropriately later on.

Author: Changed `active` to `locked` because we introduce `active` in the next paragraph with other meaning. and also `stake` -> `sub-stake` (here it's very important).
About winding down: sometimes better to say twice :) but if you sure - I can remove

        :textual:

                         Before             

            stake
            ^
            |
         900+-----------------------+
            |                       |
            |          1st          |
            |                       |   period
            +---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6       
			
			
			
                         After             

            stake
            ^
            |
         900+-----------------------+
            |                       |
            |          1st          |
         300+-----------------------+-------+
            |              2nd              |   period
            +---+---+---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
   

Slashing
^^^^^^^^
*See:* `the slashing protocol`_ 

.. _`the slashing protocol`: https://docs.nucypher.com/en/latest/architecture/slashing.html

Flags that affect the sub-stake array
-------------------------------------

Re-staking
^^^^^^^^^^
*Used in methods* : ``confirmActivity()``, ``mint()``

When re-staking is turned off, the number of locked tokens in sub-stakes does not change by itself.

> However, when re-staking is enabled (default) then all staking rewards are re-locked as part of each relevant sub-stake (inside ``confirmActivity()`` and/or ``mint()``).  Consequently, each such sub-stake has an increased locked amount (by reward) and the number of sub-stakes remains unchanged.

Reviewer 2: ```suggestion
However, when re-staking is enabled (default) then all staking rewards are re-locked as part of each relevant sub-stake (inside ``confirmActivity()`` and/or ``mint()``).  Consequently, each such sub-stake has an increased locked amount (by the accrued staking reward) and the number of sub-stakes remains unchanged.
```

Author: ✔️

            stake
            ^
            |
         700|   +-------+
            |   |  3rd  |
         600+---+-------+
         500|           +-----------+
            |    2nd    |    3rd    |
         400+-----------+-----------+-----------+
            |                                   |
            |                1st                |   period
            +---+---+---+---+---+---+---+---+---+---->    
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
			
			
			
			
                         After, re–staking is enabled             

            stake
            ^
            |
         770|   +-------+
            |   |  3rd  |
         670+---+-------+
            |           |
         550|    2nd    +-----------+
            |           |    3rd    |
         450+-----------+-----------+-----------+
            |                                   |
            |                1st                |
            |                                   |   period
            +---+---+---+---+---+---+---+---+---+---->    
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
			

Winding down
^^^^^^^^^^^^
*Used in methods* : ``confirmActivity()``

> A disabled "winding down" parameter (default) guarantees that the worker must call ``confirmActivity()`` at least N times after the parameter is enabled to unlock the sub-stake, where N is the locking duration of sub-stake. When disabled, the unlock date for each sub-stakes shifts forward by 1 period after each period i.e. the duration continues to remain the same until the parameter is enabled. Once the "winding down" parameter is enabled, each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of locking duration for each sub-stake, and the unlock date no longer changes.

Reviewer 2: Is it more intuitive to explain an enabled "winding down" parameter first?

Something like:
```suggestion
An enabled "winding down" parameter means that each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of locking duration for each sub-stake. It guarantees that the sub-stake will unlock after the worker calls ``confirm-activity()`` at least N times (no more than once in a period), where N is the locking duration of sub-stake. When disabled (default), the unlock date for each sub-stakes shifts forward by 1 period after each period i.e. the duration continues to remain the same until the parameter is enabled.
```

Author: I guess it will be better but something wrong with this particular suggestion (confused by guranties part)

Reviewer 2: ```suggestion
An enabled "winding down" parameter means that each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of locking duration for each sub-stake. In other words, the sub-stake will unlock after the worker calls ``confirm-activity()`` at least N times (no more than once in a period), where N is the locking duration of sub-stake. When disabled (default), the unlock date for each sub-stakes shifts forward by 1 period after each period. In other words, the duration continues to remain the same until the "winding down" parameter is enabled.```

Author: added!
Also if you could add somewhere that when winding down enabled - unlock date remains the same

**Example:**

    A staker has few sub-stakes, worker calls ``сonfirmActivity()`` each period:
		- Current period: 
			* 1st sub-stake = 400 tokens for 8 periods
			* 2nd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods
		- Next period, if winding down is disabled:  
			* 1st sub-stake = 400 tokens for 8 periods
			* 2nd sub-stake = 100 tokens locked starting from the current period and locked for 5 future periods
		- Next period, if winding down is enabled: 
			* 1st sub-stake = 400 tokens for 7 periods

> 			* 2nd sub-stake = 100 tokens locked starting from the current period and locked for 4 future periods

Reviewer 2: Is this right? Before this period, the 2nd sub-stake hadn't started. So there should be 5 remaining periods?

Author: It's correct, will be exactly 5 `confirmActivity()` calls (including 0 period call)

Sub-stakes
==========

A staker may extend the unlock period for any number of portions of their total stake. This divides the stake into smaller parts, each with a unique unlock date in the future. Stakers may also acquire and lock new tokens. The total stake is represented as the sum of all the different sub-stakes active in a given period (new period every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn. 

> Each sub-stake has a beginning and locking duration. For our terminology, we define `locking duration` to be the required number of times that `confirmActivity()` needs to be called so that the sub-stake is registered as locked for the subsequent period. For example, if a sub-stake has a *locking duration of 5 periods* it means that the sub-stake will be locked for periods: 1,2,3,4,5 because `confirmActivity()` would be called during periods 0,1,2,3,4 each time confirming activity for the subsequent period (1,2,3,4,5) respectively. Unless the starting period is specified, the sub-stake is considered locked for the 0th (current) period by default. When a staker confirms activity each day and winding down is enabled, the remaining lock time for relevant sub-stakes is reduced.

Reviewer: ```suggestion
Each sub-stake has a starting period and a locking duration. For our terminology, we define "locking duration" to be the required number of times that ``confirmActivity()`` needs to be called so that the sub-stake is registered as locked for the subsequent period. For example, if a sub-stake has a *locking duration of 5 periods* it means that the sub-stake will be locked for periods: 1,2,3,4,5 because ``confirmActivity()`` would be called during periods 0,1,2,3,4 each time confirming activity for the subsequent period (1,2,3,4,5) respectively. Unless the starting period is specified, the sub-stake is considered locked for the 0th (current) period by default. When a staker confirms activity each day and winding down is enabled, the remaining lock time for relevant sub-stakes is reduced.
```

Author: ✔️

Operations that modify the sub-stake array
------------------------------------------

Deposit and locking
^^^^^^^^^^^^^^^^^^^
*Methods* : ``deposit(uint256,uint16)``,  ``deposit(address,uint256,uint16)``,  ``lock(uint256,uint16)``

To become a staker, NU tokens must be transferred to the ``StakingEscrow`` contract and locked using one of the ``deposit()`` methods. If the staker already has unlocked tokens within the account in the contract, then the stake can be locked using the ``lock()`` method. If successful, a new element will be created in the array of sub-stakes with the next period as the starting date, and the duration equal to the input parameter.

**Example:**

    A staker deposits 900 tokens:

>         * 1st sub-stake = 900 tokens starting from the next period and locked for 5 periods

Reviewer: we should use our 'locking duration' term - similar comment for other spots on the page.
```suggestion
        * 1st sub-stake = 900 tokens starting from the next period and a locking duration of 5 periods
```

Author: ✔️

    .. aafig::
        :proportional:
        :textual:

            stake
            ^
            |
         900|   +-------------------+
            |   |        1st        |   period
            +---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6

Lock prolongation
^^^^^^^^^^^^^^^^^
*Methods* : ``prolongStake(uint256,uint16)``

In order to increase the staking reward, as well as the possibility of obtaining policies with a longer timeframe, stakers can increase the duration of their locked sub-stake using the ``prolongStake()`` method. The number of sub-stakes does not change, but the locked duration for the specified sub-stake will be increased.

**Example:**

    A staker prolongs sub-stake for 2 additional periods up to 7:
		- Before: 

> 			* 1st sub-stake = 900 tokens for 5 periods

Reviewer: we should use our 'locking duration' term - similar comment for other spots in the page.

```suggestion
			* 1st sub-stake = 900 tokens with locking duration of 5 periods
```

Author: ✔️

			* 2nd sub-stake = 300 tokens for 7 periods

    .. aafig::
        :proportional:
        :textual:

                         Before             

            stake
            ^
            |
         900+-----------------------+
            |                       |
            |          1st          |
            |                       |   period
            +---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6       
			
			
			
                         After             

            stake
            ^
            |
         900+-----------------------+
            |                       |
            |          1st          |
         300+-----------------------+-------+
            |              2nd              |   period
            +---+---+---+---+---+---+---+---+---->
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
   

Slashing
^^^^^^^^
*See:* `the slashing protocol`_ 

.. _`the slashing protocol`: https://docs.nucypher.com/en/latest/architecture/slashing.html

Flags that affect the sub-stake array
-------------------------------------

Re-staking
^^^^^^^^^^
*Used in methods* : ``confirmActivity()``, ``mint()``

> When re-staking is turned off, the number of locked tokens in sub-stakes does not change by itself.

Reviewer: Since we used enabled later on, we should use disabled here.
```suggestion
When re-staking is disabled, the number of locked tokens in sub-stakes does not change by itself.
```

Author: ✔️

            stake
            ^
            |
         700|   +-------+
            |   |  3rd  |
         600+---+-------+
         500|           +-----------+
            |    2nd    |    3rd    |
         400+-----------+-----------+-----------+
            |                                   |
            |                1st                |   period
            +---+---+---+---+---+---+---+---+---+---->    
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
			
			
			
			
                         After, re–staking is enabled             

            stake
            ^
            |
         770|   +-------+
            |   |  3rd  |
         670+---+-------+
            |           |
         550|    2nd    +-----------+
            |           |    3rd    |
         450+-----------+-----------+-----------+
            |                                   |
            |                1st                |
            |                                   |   period
            +---+---+---+---+---+---+---+---+---+---->    
            | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
			

Winding down
^^^^^^^^^^^^
*Used in methods* : ``confirmActivity()``

> An enabled "winding down" parameter means that each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of locking duration for each sub-stake. In other words, the sub-stake will unlock after the worker calls ``confirm-activity()`` at least N times (no more than once in a period), where N is the locking duration of sub-stake. When disabled (default), the unlock date for each sub-stakes shifts forward by 1 period after each period. In other words, the duration continues to remain the same until the "winding down" parameter is enabled.

Reviewer: ```suggestion
An enabled "winding down" parameter means that each call to ``confirmActivity()`` (no more than once in a period) leads to a reduction of the locking duration for each sub-stake. In other words, the sub-stake will unlock after the worker calls ``confirm-activity()`` at least N times (no more than once in a period), where N is the locking duration of sub-stake. When disabled (default), the unlock date for each sub-stakes shifts forward by 1 period after each period. In other words, the duration continues to remain the same until the "winding down" parameter is enabled.
```

Author: ✔️

Sub-stakes
==========

A staker may extend the unlock period for any number of portions of their total stake. This divides the stake into smaller parts, each with a unique unlock date in the future. Stakers may also acquire and lock new tokens. The total stake is represented as the sum of all the different sub-stakes active in a given period (new period every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn. 

Each sub-stake has:

    * **Locked amount of tokens**

>         The locked amount of tokens is fixed for all applicable periods i.e. it is not possible for a single sub-stake to have a different amount of tokens locked for different periods. To facilitate such a scenario, sub-stakes would need to be divided, see `Sub-stake division`_.

Reviewer 3: ```suggestion
        The locked amount of tokens is fixed for all applicable periods, i.e., it is not possible for a single sub-stake to have a different amount of tokens locked for different periods. To facilitate such a scenario, sub-stakes would need to be divided, see `Sub-stake division`_.
```

Author: ✔️

    .. aafig::
        :proportional:
        :textual:

                         Before             

            stake
            ^
            |
         900+-----------------------+
            |          1st          |   period
            +---+---+---+---+---+---+---->
            + 0 + 1 + 2 + 3 + 4 + 5 + 6      
			
			
			
                         After             

            stake
            ^
            |
         900+-------------------------------+
            |              1st              |   period
            +---+---+---+---+---+---+---+---+---->
            + 0 + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8
			
			
Sub-stake division
^^^^^^^^^^^^^^^^^^
*Methods* : ``divideStake(uint256,uint256,uint16)``

If necessary, stakers can extend the locking duration for only a portion of their tokens in a sub-stake by using the ``divideStake()`` method. This method splits a sub-stake into two elements: the old sub-stake with the reduced locked amount and the new sub-stake with the specified amount. The new sub-stake has the specified locked amount and an extended lock duration, based on the specified number of periods, with the same start period as the old sub-stake.

> 

Reviewer 3: A fundamental question for me is: why would a staker want to divide her stake? What's the use case or goal behind that? Can we give an example or motivation?

Reviewer 3: Also, substake division needs a link from the staking guide. See
https://github.com/nucypher/nucypher/commit/36204bd86cf692037dbc1bb0278be14700a50ae8

Author: This supposed to be an answer: `can extend the locking duration for only a portion of their tokens in a sub-stake`. Any suggestions how to improve?
